src/common/PopupContainerViewBase.tsx

```diff
@@ -28,6 +29,249 @@ export interface PopupComponent {
+                    if (anchorOffset < MIN_ANCHOR_OFFSET || anchorOffset > effectiveWidth - MIN_ANCHOR_OFFSET) {
+                        foundPerfectFit = false;
+                    }
+                }
+                absX = windowDims.width - ALLEY_WIDTH - effectiveWidth;
+            }
+
+            if (absY < ALLEY_WIDTH) {
+                if (pos === 'right' || pos === 'left') {
+                    anchorOffset += absY - ALLEY_WIDTH;
+                    if (anchorOffset < MIN_ANCHOR_OFFSET || anchorOffset > effectiveHeight - MIN_ANCHOR_OFFSET) {
+                        foundPerfectFit = false;
+                    }
+                }
+                absY = ALLEY_WIDTH;
+            } else if (absY > windowDims.height - ALLEY_WIDTH - effectiveHeight) {
+                if (pos === 'right' || pos === 'left') {
+                    anchorOffset -= (windowDims.height - ALLEY_WIDTH - effectiveHeight - absY);
+                    if (anchorOffset < MIN_ANCHOR_OFFSET || anchorOffset > effectiveHeight - MIN_ANCHOR_OFFSET) {
+                        foundPerfectFit = false;
+                    }
+                }
+                absY = windowDims.height - ALLEY_WIDTH - effectiveHeight;
+            }
+
+            if (foundPerfectFit || effectiveHeight > 0 || effectiveWidth > 0) {
+                result.popupY = absY;
+                result.popupX = absX;
+                result.anchorOffset = anchorOffset;
+                result.anchorPosition = pos;
+                result.constrainedPopupWidth = effectiveWidth;
+                result.constrainedPopupHeight = effectiveHeight;
+
+                foundPartialFit = true;
+            }
+        }
+    });
+
+    return result;
+}
+
+function recalcInnerPosition(anchorRect: ClientRect, positionToUse: PopupPosition, popupWidth: number, popupHeight: number) {
+    // For inner popups we only accept the first position of the priorities since there
+    // should always be room for the bubble.
+    let popupX = 0;
+    let popupY = 0;
+    let anchorOffset = 0;
+    switch (positionToUse) {
+        case 'top':
+            popupY = anchorRect.top + anchorRect.height - popupHeight;
+            popupX = anchorRect.left + anchorRect.height / 2 - popupWidth / 2;
+            anchorOffset = popupWidth / 2;
+            break;
+
+        case 'bottom':
+            popupY = anchorRect.top + popupHeight;
+            popupX = anchorRect.left + anchorRect.height / 2 - popupWidth / 2;
+            anchorOffset = popupWidth / 2;
+            break;
+
+        case 'left':
+            popupY = anchorRect.top + anchorRect.height / 2 - popupHeight / 2;
+            popupX = anchorRect.left + anchorRect.width - popupWidth;
+            anchorOffset = popupHeight / 2;
+            break;
+
+        case 'right':
+            popupY = anchorRect.top + anchorRect.height / 2 - popupHeight / 2;
+            popupX = anchorRect.left;
+            anchorOffset = popupHeight / 2;
+            break;
+    }
```

Comment:
This case doesn't handle the new 'context' value. I presume it's considered a programming error to specify 'context' with inner position. We should make sure this is documented and doesn't crash.